director/prom_query.go

```diff
@@ -0,0 +1,266 @@
+/***************************************************************
+ *
+ * Copyright (C) 2024, Pelican Project, Morgridge Institute for Research
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ***************************************************************/
+
+package director
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/pelicanplatform/pelican/config"
+	"github.com/pelicanplatform/pelican/param"
+	"github.com/pelicanplatform/pelican/token"
+	"github.com/pelicanplatform/pelican/token_scopes"
+)
+
+type (
+	promQLResData struct {
+		ResultType string // "matrix" | "vector" | "scalar" | "string"
+
+		// Result has varying formats depending on the ResultType.
+		// See the [expression query result formats]: https://prometheus.io/docs/prometheus/latest/querying/api/#expression-query-result-formats.
+		Result []interface{}
+	}
+	promQLRes struct {
+		Status    string // success | error
+		Data      promQLResData
+		Error     string // only when status == "error"
+		ErrorType string // only when status == "error"
+	}
+
+	promQLValuePair struct {
+		UnixTime float64
+		// Numbers are encoded in string to include NaN, Inf, and -Inf.
+		// It is the user's responsibility to handle the conversion from string to numbers.
+		Value string
+	}
+
+	promQLResultItem struct {
+		Metric map[string]interface{}
+		Values []promQLValuePair
+	}
+
+	promQLParsed struct {
+		ResultType string // "matrix" | "vector" | "scalar" | "string"
+		Result     []promQLResultItem
+	}
+)
+
+// Query the Prometheus PromQL endpoint on the director server
+// at /api/v1.0/prometheus/query?query=
+//
+// where the only arg is the query to execute, without "?query="
+//
+// Example: queryPromtheus("up") // Get metric of the running Prometheus instances
+func queryPromtheus(ctx context.Context, query string, withToken bool) (promParsed promQLParsed, err error) {
+	if strings.HasPrefix(query, "?query=") {
+		err = errors.Errorf("query argument should not contain \"?query=\"")
+		return
+	}
+	extWebUrl := param.Server_ExternalWebUrl.GetString()
+	baseUrl, err := url.JoinPath(extWebUrl, "/api/v1.0/prometheus/query")
+	if err != nil {
+		return
+	}
+	queryUrl := baseUrl + "?query=" + url.QueryEscape(query)
```

Comment:
Why not construct this so that if the incoming query is already prefixed with `?query=` we simply pass it along, but if there is no prefix, we prepend it?

Reply:
Oh that's because we can't tell if someone is mistakenly passing ?query=xxx where xxx is not URLEncoded, so we require that they only pass the query body instead of with the prefix